### PR TITLE
patch up exp backoff; pre rm_rf on fetch

### PR DESCRIPTION
**Pull request by kalefranz**
CC @mingwandroid @mcg1969

### CODE REVIEW

**Comment by kalefranz** [comment]
+ CC @ilanschnell

File: conda/install.py [hunk 1]
```diff
@@ -227,17 +227,20 @@ def exp_backoff_fn(fn, *args):
...
+    import random
     max_retries = 5
     for n in range(max_retries):
         try:
             result = fn(*args)
-        except Exception as e:
+        except EnvironmentError as e:
             log.debug(repr(e))
-            if n == max_retries-1:
+            if e.errno in (errno.EPERM, errno.EACCES):
```
Comment (kalefranz) [inline conda/install.py]: @mingwandroid can you help me verify that these are the right errors to catch here?

File: conda/install.py [hunk 2]
```diff
@@ -227,17 +227,20 @@ def exp_backoff_fn(fn, *args):
...
         return fn(*args)
-    import random
 
+    import random
     max_retries = 5
     for n in range(max_retries):
         try:
             result = fn(*args)
-        except Exception as e:
+        except EnvironmentError as e:
```
Comment (mingwandroid) [inline conda/install.py]: `PermissionError` in one case, the other I couldn't get to repeat, but hopefully the same.

File: conda/install.py [hunk 3]
```diff
@@ -227,17 +227,20 @@ def exp_backoff_fn(fn, *args):
...
             log.debug(repr(e))
-            if n == max_retries-1:
+            if e.errno in (errno.EPERM, errno.EACCES):
+                if n == max_retries-1:
+                    raise
+                time.sleep(((2 ** n) + random.random()) * 1e-3)
```
Comment (mingwandroid) [inline conda/install.py]: Here I'd actually like, when debugging, enough time to go to the file in question using Unlocker to see what's got it open, perhaps we can have it in seconds for debug?

Comment (kalefranz) [inline conda/install.py]: Let's add a config parameter for that, which I'll do in 4.2 when I redo configs.  Deal?

File: conda/install.py (hunk 2 above)
Comment (kalefranz) [inline conda/install.py]: The goal here was to have the most general error that assigns errno.  But apparently it's changed in python 3

> Changed in version 3.3: EnvironmentError, IOError, WindowsError, socket.error, select.error and mmap.error have been merged into OSError, and the constructor may return a subclass.

https://docs.python.org/3/library/exceptions.html#OSError.filename2

So I think `OSError` is the safe choice here?  As long as neither `errno.EPERM` or `errno.EACCES` is an `IOError` in python2.7

Comment (kalefranz) [inline conda/install.py]: https://docs.python.org/3/library/exceptions.html#PermissionError

> exception PermissionError¶
> Raised when trying to run an operation without the adequate access rights - for example filesystem permissions. Corresponds to errno EACCES and EPERM.

File: conda/install.py (hunk 3 above)
Comment (mingwandroid) [inline conda/install.py]: Sure thing. One final matter, as @ilanschnell pointed out, Python 2 needs:

```from __future__ import division```

Comment (kalefranz) [inline conda/install.py]: It was already there.

Comment (mingwandroid) [inline conda/install.py]: Oops, yeah, of course.

**Pull request merged**